torch_np/_ufuncs.py

```diff
@@ -133,4 +134,64 @@ def modf(x, /, *args, **kwds):
     return rem, quot
 
 
-__all__ = __all__ + ["divmod", "modf", "matmul"]
+__binary__ = __binary__ + ["divmod", "modf", "matmul"]
+
+
+# ############# Unary ufuncs ######################
+
+
+__unary__ = [
+    name
+    for name in dir(_unary_ufuncs_impl)
+    if not name.startswith("_") and name != "torch"
+]
+
+
+def deco_unary_ufunc(torch_func):
+    """Common infra for unary ufuncs.
+
+    Normalize arguments, sort out type casting, broadcasting and delegate to
+    the pytorch functions for the actual work.
+    """
+
+    def wrapped(
+        x: ArrayLike,
+        /,
+        out: Optional[NDArray] = None,
+        *,
+        where=True,
+        casting="same_kind",
+        order="K",
+        dtype: DTypeLike = None,
+        subok: SubokLike = False,
+        signature=None,
+        extobj=None,
+    ):
+        tensors = _helpers.ufunc_preprocess(
+            (x,), out, where, casting, order, dtype, subok, signature, extobj
+        )
+        # now broadcast the input tensor against the out=... array
+        if out is not None:
+            # XXX: need to filter out noop broadcasts if t.shape == out.shape?
+            shape = out.shape
+            tensors = tuple(torch.broadcast_to(t, shape) for t in tensors)
+        result = torch_func(*tensors)
+        return result
+
+    return wrapped
+
+
+#
+# For each torch ufunc implementation, decorate and attach the decorated name
+# to this module. Its contents is then exported to the public namespace in __init__.py
+#
+for name in __unary__:
+    ufunc = getattr(_unary_ufuncs_impl, name)
+    decorated = normalizer(deco_unary_ufunc(ufunc))
+
+    decorated.__qualname__ = name  # XXX: is this really correct?
+    decorated.__name__ = name
```

Comment:
We're not using `@wraps` here. This decorator actually defines the signature and wraps a pytorch function which does not have numpy ufunc arguments.

The `__module__` attribute is set correctly automagically somehow though,

```
In [3]: tnp.add.__module__
Out[3]: 'torch_np._ufuncs'
```

What I'm not sure about is what the `__qualname__` is and what it should be. E.g., with NumPy itself,

```
In [4]: import numpy as np

In [5]: np.add.__name__
Out[5]: 'add'

In [6]: np.add.__qualname__
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 np.add.__qualname__

AttributeError: 'numpy.ufunc' object has no attribute '__qualname__'

In [7]: np.array.__qualname__
Out[7]: 'array'

```

Reply:
We are using `wraps` at
https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/3ec9ddae691894d1f9534ef68c2be8e23bcf5384/torch_np/_normalizations.py#L141-L142
`__qualname__` stands for "qualified name". It means that it returns the full qualified name of a given function. For example, for a method `foo` of a `class A`, it'll return `A.foo`, while if `foo` is a free function, it'll return just `foo`. 

This reminds me, we should probably set this variable to its correct value when registering the methods of `torch_np.ndarray`. Same for their name, which is not correctly set as we're just assigning to them. Now, this is fairly low prio, so I wouldn't worry too much about it for now.

Reply:
Removing these assignments gives

```
In [1]: import torch_np as tnp

In [2]: tnp.sin.__qualname__
Out[2]: 'deco_unary_ufunc.<locals>.wrapped'

In [3]: tnp.sin.__name__
Out[3]: 'wrapped'
```

which is precisely why they were added in the first place. It's full well possible that this is because I'm doing something stupid somewhere in the decorator chain, but I'm not seeing what it is.

Reply:
You are missing a `functools.wraps` in
https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/3ec9ddae691894d1f9534ef68c2be8e23bcf5384/torch_np/_unary_ufuncs.py#L14-L21
Probably also for the `deco_binary_ufunc`, I didn't check.

Reply:
Like I'm saying, it's deliberate. The signature is in `wrapped`, not `torch_func`.

Reply:
I'm pretty sure that `functools.wraps` handles the change of signature just fine.

Reply:
OK, let me summarize an off-line discussion with @lezcano.
Consider `_ufuncs_impl.invert`: 
- it's `from torch import bitwise_not as invert`: the rename is give it a numpy-compatible name.
- if we propagate its attributes via `@wraps(torch_func)`, we get the  `__qualname__` of `torch.bitwise_not`, which is not what we're after.

We can attach the `__name__` in the decorator, not in a module-level loop  (FWIW I don't see a point, but that's irrelevant; if there at least one reason to, fine). But what is the `__qualname__` for a module-level function `torch_np.invert`, which is implemented in `torch_np._ufuncs` module? @lezcano 

As a datapoint, `numpy.invert` simply does not have `__qualname__`

Reply:
OK, done in https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/105/commits/14273c8987213174b919cff037fe60b3aa6862df

This does have a side effect that implementation details leak into the names but if that's wanted, fine :-).

```
In [1]: import torch_np as tnp

In [2]: tnp.invert.__name__
Out[2]: 'bitwise_not'
```

Reply:
That was also an issue with the previous implementation. The way to fix that is by rather than doing `from torch import bitwise_not as invert`, you should have a function that registers a given function with the correct name. See e.g.
https://github.com/pytorch/pytorch/blob/37b9143206dbe53aa27405a0a889b2fd9c76ec8b/torch/_refs/__init__.py#L416-L426

Reply:
It wasn't because the previous implementation never used `torch_func.__name__`. It used 
 `decorated.__name__ = name` , where `name` came from `dir(_ufunc_impl)`, so it picked  `from torch import bitwise_not as invert`. 

Taking a step back, I do not think it's a big deal and the status quo is good enough. WDYT?